File: cplusplus/746MinCostClimbingStairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        if (cost.size() <= 1)
            return 0;
        
        vector<int> totalCost(cost.size() + 1, 0);
        totalCost[0] = 0;
        totalCost[1] = 0;
        
        for (int i = 2; i <= cost.size(); ++i) {
            totalCost[i] = totalCost[i-1] + cost[i-1];
            if (totalCost[i] > totalCost[i-2] + cost[i-2])
                totalCost[i] = totalCost[i-2] + cost[i-2];
        }
        
        return totalCost[cost.size()];
    }
};

File: java/746MinCostClimbingStairs.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        if (cost.length == 0)
            return 0;
        int[] minCost = new int[cost.length];
        minCost[cost.length - 1] = cost[cost.length - 1];
        if (cost.length == 0)
            return minCost[0];
        
        minCost[cost.length - 2] = Math.min(cost[cost.length - 2], cost[cost.length - 1] + cost[cost.length - 2]);
        for (int i = cost.length - 3; i >= 0; --i) {
            minCost[i] = cost[i] + Math.min(minCost[i+1], minCost[i+2]);
        }
        
        return Math.min(minCost[0], minCost[1]);
    }
}
